zebra/debug_nl.c

```diff
@@ -0,0 +1,1291 @@
+		return;
+
+	plen = RTA_PAYLOAD(rta);
+	zlog_debug("    rta [len=%d (payload=%lu) type=(%d) %s]", rta->rta_len,
+		   plen, rta->rta_type, rta_type2str(rta->rta_type));
+	switch (rta->rta_type) {
+	case IFLA_IFNAME:
+	case IFLA_IFALIAS:
+		if (plen < sizeof(char)) {
+			zlog_debug("      invalid length");
+			break;
+		}
+
+		snprintf(dbuf, plen, "%s", (char *)RTA_DATA(rta));
+		zlog_debug("      %s", dbuf);
+		break;
+
+	case IFLA_MTU:
+	case IFLA_TXQLEN:
+	case IFLA_NUM_TX_QUEUES:
+	case IFLA_NUM_RX_QUEUES:
+	case IFLA_GROUP:
+	case IFLA_PROMISCUITY:
+#ifdef IFLA_GSO_MAX_SEGS
+	case IFLA_GSO_MAX_SEGS:
+#endif /* IFLA_GSO_MAX_SEGS */
+#ifdef IFLA_GSO_MAX_SIZE
+	case IFLA_GSO_MAX_SIZE:
+#endif /* IFLA_GSO_MAX_SIZE */
+	case IFLA_CARRIER_CHANGES:
+	case IFLA_MASTER:
+		if (plen < sizeof(uint32_t)) {
+			zlog_debug("      invalid length");
+			break;
+		}
+
+		u32v = *(uint32_t *)RTA_DATA(rta);
+		zlog_debug("      %u", u32v);
+		break;
+
+	case IFLA_ADDRESS:
+		datap = RTA_DATA(rta);
+		dbufpos = 0;
+		dbuf[0] = 0;
+		for (it = 0; it < (int)plen; it++) {
+			wlen = snprintf(dbuf + dbufpos, sizeof(dbuf) - dbufpos,
+					"%02X:", *datap);
+			dbufpos += (size_t)wlen;
+			datap++;
+		}
+		/* Remove trailing ':'. */
+		dbuf[dbufpos - 1] = 0;
+
+		if (dbuf[0])
+			zlog_debug("      %s", dbuf);
+		else
+			zlog_debug("      <empty>");
+		break;
+
+	case IFLA_LINKINFO:
+		nllink_linkinfo_dump(RTA_DATA(rta), msglen);
+		break;
+
+	default:
+		/* NOTHING: unhandled. */
+		break;
+	}
+
+	/* Get next pointer and start iteration again. */
+	rta = RTA_NEXT(rta, msglen);
+	goto next_rta;
+}
+
+static void nlroute_dump(struct rtmsg *rtm, size_t msglen)
+{
+	struct rtattr *rta;
+	size_t plen;
+	uint32_t u32v;
+	char nbuf[128];
+
+	/* Get the first attribute and go from there. */
+	rta = RTM_RTA(rtm);
+next_rta:
+	/* Check the header for valid length and for outbound access. */
+	if (RTA_OK(rta, msglen) == 0)
+		return;
+
+	plen = RTA_PAYLOAD(rta);
+	zlog_debug("    rta [len=%d (payload=%lu) type=(%d) %s]", rta->rta_len,
+		   plen, rta->rta_type, rtm_rta2str(rta->rta_type));
+	switch (rta->rta_type) {
+	case RTA_IIF:
+	case RTA_OIF:
+	case RTA_PRIORITY:
+	case RTA_TABLE:
+	case RTA_NH_ID:
+		u32v = *(uint32_t *)RTA_DATA(rta);
+		zlog_debug("      %u", u32v);
+		break;
+
+	case RTA_GATEWAY:
+	case RTA_DST:
+	case RTA_SRC:
+	case RTA_PREFSRC:
+		switch (plen) {
+		case 4:
+			zlog_debug("      %s", inet_ntop(AF_INET, RTA_DATA(rta),
+							 nbuf, sizeof(nbuf)));
+			break;
+		case 16:
+			zlog_debug("      %s",
+				   inet_ntop(AF_INET6, RTA_DATA(rta), nbuf,
+					     sizeof(nbuf)));
+			break;
+		default:
+			break;
+		}
+		break;
+
+	default:
+		/* NOTHING: unhandled. */
+		break;
+	}
+
+	/* Get next pointer and start iteration again. */
+	rta = RTA_NEXT(rta, msglen);
+	goto next_rta;
+}
+
+static void nlneigh_dump(struct ndmsg *ndm, size_t msglen)
+{
+	struct rtattr *rta;
+	uint8_t *datap;
+	size_t dbufpos;
+	size_t plen;
+	int it, wlen;
```

Comment:
make `it` a `size_t` and avoid value-changing cast later?

Reply:
Fixed.